### PR TITLE
Pkt ptrs/v39

### PR DESCRIPTION
**Pull request by victorjulien**
Redo much of the header pointer logic, cleaning up macro's and unifying various pointers that are mutually exclusive into unions.

Reduces size of the Packet struct from almost 600 bytes to around 450 bytes.

Includes https://github.com/OISF/suricata/pull/10762
Replaces https://github.com/OISF/suricata/pull/10791

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 19989

**Comment by catenacyber** [comment]
Is there a redmine ticket for this ?

**Review by catenacyber — COMMENTED** [review]
Thanks for the work Victor :-)

- CI : 🟢 
- Code : checking now
- Commits segmentation : looks ok
- Commit messages : do we need tickets ?
One line with `For the tests, switch to getting headers from embedded IPv6 header.` is a bit too long and may be its own and only line.
- Git ID set : looks fine for me
- CLA : you already contributed :-p
- Doc update : not needed
- Redmine ticket : 🟠 to do ?
- Rustfmt : not needed
- Tests : how do you compute the size of the Packet struct ?
- Dependencies added: none

**Comment by catenacyber** [comment]
> decode: implement IP_GET_IPPROTO as inline func

Why ? Code readability ?

**Comment by catenacyber** [comment]
> decode: turn PKT_IS_IPV4/PKT_IS_IPV6 into functions

Bad commit title: this creates a rapport, but `PKT_IS_IPV4` still exists as a macro and is used as such by PacketIsIPv4

**Comment by catenacyber** [comment]
@jasonish what do you think about the commit 15c2da963aa2b79800ad0832488513bb240d76ab ? It is about defrag

File: src/util-validate.h [hunk 1]
```diff
@@ -65,33 +65,32 @@
...
-            BUG_ON((p)->payload == NULL);           \
-        }                                           \
-        BUG_ON((p)->ip4h != NULL && (p)->ip6h != NULL);     \
-        BUG_ON((p)->flowflags != 0 && (p)->flow == NULL);   \
-        BUG_ON((p)->flowflags & FLOW_PKT_TOSERVER &&\
-               (p)->flowflags & FLOW_PKT_TOCLIENT); \
-    }                                               \
-} while(0)
+#define DEBUG_VALIDATE_PACKET(p)                                                                   \
+    do {                                                                                           \
```
Comment (catenacyber) [inline src/util-validate.h:69]: Why did you add a do while wrapper ?

**Comment by catenacyber** [comment]
> decode: clean up pointer list

Not a good title, would prefer something like `style cleanup` to say that this does not change code

File: src/util-validate.h (hunk 1 above)
Comment (victorjulien) [reply src/util-validate.h:69]: many macro's have it, but this commit didn't add it

File: src/decode-tcp.c [hunk 2]
```diff
@@ -89,47 +90,54 @@ static void DecodeTCPOptions(Packet *p, const uint8_t *pkt, uint16_t pktlen)
...
                     break;
                 case TCP_OPT_SACKOK:
                     if (olen != TCP_OPT_SACKOK_LEN) {
                         ENGINE_SET_EVENT(p,TCP_OPT_INVALID_LEN);
                     } else {
-                        if (p->tcpvars.sackok.type != 0) {
+                        if (TCP_GET_SACKOK(p)) {
                             ENGINE_SET_EVENT(p,TCP_OPT_DUPLICATE);
                         } else {
-                            SET_OPTS(p->tcpvars.sackok, tcp_opts[tcp_opt_cnt]);
```
Comment (catenacyber) [inline src/decode-tcp.c:117]: Looks like you can get rid of `SET_OPTS` definition

Comment (catenacyber) [reply src/decode-tcp.c:117]: Because it is unused after your chan

**Comment by victorjulien** [comment]
> > decode: implement IP_GET_IPPROTO as inline func
> 
> Why ? Code readability ?

Yeah, mostly. Plus the type checking. But the many and often nested macro calls have often made code hard to follow.

File: src/decode-tcp.c [hunk 3]
```diff
@@ -89,47 +90,54 @@ static void DecodeTCPOptions(Packet *p, const uint8_t *pkt, uint16_t pktlen)
...
                     if (olen != TCP_OPT_MSS_LEN) {
                         ENGINE_SET_EVENT(p,TCP_OPT_INVALID_LEN);
                     } else {
-                        if (p->tcpvars.mss.type != 0) {
+                        if (p->l4.vars.tcp.mss_set) {
                             ENGINE_SET_EVENT(p,TCP_OPT_DUPLICATE);
                         } else {
-                            SET_OPTS(p->tcpvars.mss, tcp_opts[tcp_opt_cnt]);
+                            p->l4.vars.tcp.mss_set = true;
+                            p->l4.vars.tcp.mss = SCNtohs(*(uint16_t *)(tcp_opts[tcp_opt_cnt].data));
```
Comment (catenacyber) [inline src/decode-tcp.c:114]: Looks like this misses a check against buffer overread (if olen == 3)

**Comment by victorjulien** [comment]
> Is there a redmine ticket for this ?

Just added https://redmine.openinfosecfoundation.org/issues/6938

File: src/decode-tcp.c [hunk 4]
```diff
@@ -89,47 +90,54 @@ static void DecodeTCPOptions(Packet *p, const uint8_t *pkt, uint16_t pktlen)
...
+                            if (wscale <= TCP_WSCALE_MAX) {
+                                p->l4.vars.tcp.wscale = wscale;
+                            } else {
+                                p->l4.vars.tcp.wscale = 0;
+                            }
                         }
                     }
                     break;
                 case TCP_OPT_MSS:
                     if (olen != TCP_OPT_MSS_LEN) {
```
Comment (victorjulien) [inline src/decode-tcp.c:107]: we check for olen != 4 here, so the code below is only reachable for `olen == 4`

File: src/decode-tcp.c (hunk 3 above)
Comment (victorjulien) [reply src/decode-tcp.c:114]: we can get here only in case of `olen == 4`, see https://github.com/OISF/suricata/pull/10796/files/2acf8dfdc0db8032c744a388c9dfe97b21f51af3#r1561112199

File: src/decode-tcp.c (hunk 2 above)
Comment (victorjulien) [reply src/decode-tcp.c:117]: Good catch

File: src/decode-tcp.h [hunk 5]
```diff
@@ -158,22 +165,20 @@ typedef struct TCPVars_
     bool md5_option_present;
     bool ao_option_present;
     bool ts_set;
+    bool sack_ok;
+    bool mss_set;
+    bool tfo_set;
+    uint8_t wscale_set : 1;
```
Comment (catenacyber) [inline src/decode-tcp.h:171]: why not a bool ?

File: src/decode-tcp.c [hunk 6]
```diff
@@ -142,10 +150,14 @@ static void DecodeTCPOptions(Packet *p, const uint8_t *pkt, uint16_t pktlen)
                     {
                         ENGINE_SET_EVENT(p,TCP_OPT_INVALID_LEN);
                     } else {
-                        if (p->tcpvars.sack.type != 0) {
+                        if (p->l4.vars.tcp.sack_set) {
                             ENGINE_SET_EVENT(p,TCP_OPT_DUPLICATE);
                         } else {
-                            SET_OPTS(p->tcpvars.sack, tcp_opts[tcp_opt_cnt]);
+                            ptrdiff_t diff = tcp_opts[tcp_opt_cnt].data - (uint8_t *)tcph;
```
Comment (catenacyber) [inline src/decode-tcp.c:156]: This is UB if olen == 2 ?

Comment (victorjulien) [reply src/decode-tcp.c:156]: what would be the UB?

File: src/defrag.c [hunk 7]
```diff
@@ -308,12 +310,12 @@ Defrag4Reassemble(ThreadVars *tv, DefragTracker *tracker, Packet *p)
...
 
             hlen = frag->hlen;
-            ip_hdr_offset = frag->ip_hdr_offset;
+            ip_hdr_offset = tracker->ip_hdr_offset;
 
             /* This is the start of the fragmentable portion of the
              * first packet.  All fragment offsets are relative to
              * this. */
-            fragmentable_offset = frag->ip_hdr_offset + frag->hlen;
+            fragmentable_offset = tracker->ip_hdr_offset + frag->hlen;
```
Comment (jasonish) [inline src/defrag.c:318]: Need to look closer, but could this be a possible behaviour change not covered by a unit test?  2 overlapping packets with offset 0.  Instead of looking at the packet for each fragment, this will look at what was last set in the tracker.

File: src/decode-tcp.h (hunk 5 above)
Comment (victorjulien) [reply src/decode-tcp.h:171]: I've made all of them bitfields for max compactness

**Comment by victorjulien** [comment]
Comments addressed in #10926.
